src/integrations/AmqplibIntegration.ts

```diff
@@ -0,0 +1,258 @@
+import Integration from './Integration';
+
+import {
+  SpanTags,
+  SpanTypes,
+  DomainNames,
+  AMQPTags,
+  ClassNames,
+} from '../Constants';
+
+import ThundraLogger from '../ThundraLogger';
+import ModuleUtils from '../utils/ModuleUtils';
+import ThundraChaosError from '../error/ThundraChaosError';
+import ExecutionContextManager from '../context/ExecutionContextManager';
+import * as opentracing from 'opentracing';
+
+const shimmer = require('shimmer');
+
+const MODULE_NAME = 'amqplib/lib/channel.js';
+const MODULE_VERSION = '>=0.5';
+
+/**
+ * {@link Integration} implementation for AMQPLIB Integration
+ * through {@code amqplib} library
+ */
+class AMQPLIBIntegration implements Integration {
+  config: any;
+  private instrumentContext: any;
+
+  constructor(config: any) {
+    ThundraLogger.debug('<AMQPLIBIntegration> Activating AMQPLIB Integration');
+
+    this.config = config || {};
+    this.instrumentContext = ModuleUtils.instrument(
+      [MODULE_NAME],
+      MODULE_VERSION,
+      (lib: any, cfg: any) => {
+        this.wrap.call(this, lib, cfg);
+      },
+      (lib: any, cfg: any) => {
+        this.doUnwrap.call(this, lib);
+      },
+      this.config,
+    );
+  }
+
+  /**
+   * @inheritDoc
+   */
+  wrap(lib: any, config: any) {
+    ThundraLogger.debug('<AMQPLIBIntegration> Wrap');
+
+    const integration = this;
+
+    function wrapSendMessage(sendMessage: Function) {
+      /**
+       * Wrap the sendMessage
+       * @param args sendMessage function parameters in order fields, properties and content.
+       */
+      return function sendMessageWithTrace(...args: any) {
+        ThundraLogger.debug(
+          `<AMQPLIBIntegration> Tracing sendMessage args: ${args}`,
+        );
+        const method = 'basic.publish';
+        const [fields, properties, content] = args;
+        const { tracer } = ExecutionContextManager.get();
+        const parentSpan = tracer.getActiveSpan();
+        const span = tracer._startSpan(fields.routingKey, {
+          childOf: parentSpan,
+          domainName: DomainNames.MESSAGING,
+          className: ClassNames.AMQP,
+          disableActiveStart: true,
+        });
+        integration.handleTags(this, config, span, method, fields);
+        span._initialized();
+        tracer.inject(span.spanContext, opentracing.FORMAT_TEXT_MAP, properties.headers);
+        try {
+          return sendMessage.apply(this, args);
+        } catch (error) {
+          ThundraLogger.error(
+            '<AMQPLIBIntegration> Error occurred while tracing AMQP ${method} method:',
+            error,
+          );
+
+          if (span) {
+            ThundraLogger.debug(
+              `<AMQPLIBIntegration> Because of error, closing AMQP span with name ${span.getOperationName()}`,
+            );
+            span.finish();
+          }
+
+          if (error instanceof ThundraChaosError) {
+            throw error;
+          } else {
+            return sendMessage.apply(this, args);
```

Comment:
We should not call original method here. Because when the code execution comes here, this means that `sendMessage.apply(this, args);` has failed above. So if we call it again here in case of error, this means that we are trying to send message twice. So, lets just throw error here

Reply:
In overall, I mean:
```js
    function wrapSendMessage(sendMessage: Function) {
      /**
       * Wrap the sendMessage
       * @param args sendMessage function parameters in order fields, properties and content.
       */
      return function sendMessageWithTrace(...args: any) {
        ThundraLogger.debug(
          `<AMQPLIBIntegration> Tracing sendMessage args: ${args}`,
        );
        let span;
        try {
          const {tracer} = ExecutionContextManager.get();

          if (!tracer) {
            ThundraLogger.debug('<AMQPLIBIntegration> Skipped tracing command as no tracer is available');
            return sendMessage.apply(this, args);
          }

          const method = 'basic.publish';
          const [fields, properties, content] = args;
          const parentSpan = tracer.getActiveSpan();
          span = tracer._startSpan(fields.routingKey, {
            childOf: parentSpan,
            domainName: DomainNames.MESSAGING,
            className: ClassNames.AMQP,
            disableActiveStart: true,
          });
          integration.handleTags(this, config, span, method, fields);
          span._initialized();
          tracer.inject(span.spanContext, opentracing.FORMAT_TEXT_MAP, properties.headers);
        } catch (error) {
          ThundraLogger.error(
              '<AMQPLIBIntegration> Error occurred while tracing AMQP ${method} method:',
              error,
          );
          if (span) {
            ThundraLogger.debug(
                `<AMQPLIBIntegration> Because of error, closing AMQP span with name ${span.getOperationName()}`,
            );
            span.setErrorTag(error);
            span.close();
            span = null;
          }
          if (error instanceof ThundraChaosError) {
            throw error;
          }
        }
        try {
          return sendMessage.apply(this, args);
        } catch (error) {
          if (span) {
            span.setErrorTag(error);
          }
        } finally {
          if (span) {
            span.close();
          }
        }
      };
    }
```